agents/sorenson_dlm/dlm_agent.py

```diff
@@ -0,0 +1,259 @@
+        previous command to complete to being the next
+        """
+        msg = str(cmd) + ';OPC?\r\n'
+        self.comm.send(msg.encode('ASCII'))
+
+    def rec_msg(self):
+        """
+        Waits for a message from the DLM, typically as a result of a query,
+        and returns it
+        """
+        dataStr = self.comm.recv(1024).decode('ASCII')
+        return dataStr.strip()
+
+    def set_overv_prot(self, voltage):
+        """
+        Sets the overvoltage protection
+        """
+        self.send_msg('*CLS')
+        self.send_msg('*RST')
+        self.send_msg('SOUR:VOLT:PROT {}'.format(voltage))
+
+        self.send_msg('SOUR:VOLT:PROT?')
+        ovp = self.rec_msg()
+        if ovp != voltage:
+            print("Error: Over voltage protection not set to requested value")
+            return
+
+        self.send_msg('STAT:PROT:ENABLE 8')
+        self.send_msg('STAT:PROT:ENABLE?')
+        enb = self.rec_msg()
+        if enb != '8':
+            print('Error: Over voltage protection failed to enable')
+            return
+
+        self.send_msg('STAT:PROT:EVENT?')
+        event = self.rec_msg()
+        if event != '0':
+            print('Error: Over voltage already tripped')
+            return
+
+    def read_voltage(self):
+        """
+        Reads output  voltage
+        """
+        self.send_msg('SOUR:VOLT?')
+        msg = self.rec_msg()
+        return msg
+
+    def read_current(self):
+        """
+        Reads output current
+        """
+        self.send_msg('MEAS:CURR?')
+        msg = self.rec_msg()
+        return msg
+
+    def sys_err_check(self):
+        """
+        Queries sytem error and returns error byte
+        """
+        self.send_msg('SYST:ERR?')
+        msg = self.rec_msg()
+        return msg
+
+
+class DLMAgent:
+    '''
+    TO DO:
+    set_current_and_voltage (simulatneously set current and voltage,
+    not sure if needed)
+
+    '''
+
+    def __init__(self, agent, ip_address , port, f_sample=2.5):
+        self.active = True
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+        self.f_sample = f_sample
+        self.take_data = False
+        self.over_volt = 0
+        self.dlm = DLM(ip_address, int(port))
+        agg_params = {'frame length': 60, }
+        self.agent.register_feed('voltages',
+                                 record=True,
+                                 agg_params=agg_params,
+                                 buffer_time=1)
+
+    def start_acq(self, session, params=None):
+        '''
+        Get voltage and current values from the sorenson, publishes them to
+        the feed
+
+        Args:
+            sampling_frequency: defaults to 2.5Hz
+        '''
+        if params is None:
+            params = {}
+
+        f_sample = params.get('sampling_frequency')
+        if f_sample is None:
+            f_sample = self.f_sample
+
+        sleep_time = 1./f_sample - 0.01
```

Comment:
There's a tool in ocs to help accomplish this called the [Pacemaker](https://ocs.readthedocs.io/en/develop/api.html?highlight=pacemaker#ocs.ocs_twisted.Pacemaker) that you might consider using.

You can see it in use in a couple agents, including the weather monitor: https://github.com/simonsobs/socs/blob/c1d9c1c77d329fc44f106c14d9fb77139781ff95/agents/vantagePro2_agent/vantage_pro2_agent.py#L123

Reply:
Added, please review the usage

Reply:
Adding the pacemaker looks alright. I'd probably put the `pm.sleep()` down where you have `time.sleep(wait_time)`, i.e. sleep after you've taken a measurement, rather than before. And then you can remove your `time.sleep(wait_time)`, that's what the pacemaker is doing. As is you're waiting twice as long as you need to for the desired sampling frequency.